_posts/2023-04-24-adopting-apache-arrow-at-cloudquery.md

```diff
@@ -0,0 +1,65 @@
+---
+layout: post
+title: "Adopting Apache Arrow at CloudQuery"
+date: "2023-04-26 00:00:00"
+author: Yevgeny Pats
+categories: [application]
+---
+<!--
+{% comment %}
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to you under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+{% endcomment %}
+-->
+
+This post is a collaboration with CloudQuery and cross-posted on the CloudQuery [blog](https://cloudquery.io/blog/adopting-apache-arrow-at-cloudquery).
+
+[CloudQuery](https://github.com/cloudquery/cloudquery) is an open source high performance ELT framework written in Go. We [previously](https://www.cloudquery.io/blog/building-cloudquery) discussed some of the architecture and design decisions that we took to build a performant ELT framework. A type system is a key component for creating a performant and scalable ELT framework where sources and destinations are decoupled. In this blog post we will go through why we decided to adopt Apache Arrow as our type system and replace our in-house implementation.
+
+# What is a Type System?
+
+Let’s quickly [recap](https://www.cloudquery.io/blog/building-cloudquery#type-system) what a type system is and why an ELT framework needs one. At a very high level, an ELT framework extracts data from some source and moves it to some destination with a specific schema.
+
+```text
+API ---> [Source Plugin]  ----->    [Destination Plugin]
+                          ----->    [Destination Plugin]
+                           gRPC
+```
+
+
+Sources and destinations are decoupled and communicate via gRPC. This is crucial to allowing the addition of new destinations and updating old destinations without requiring updates to source plugin code (which otherwise would introduce an unmaintainable architecture).
```

Comment:
So, if I understand correctly, another way of saying this is:

> Sources and destinations each have their own way to represent the schema of some dataset. In order to avoid having to write a conversion between every pair of source and destination (which would be difficult to maintain), we choose an intermediate schema representation that all sources and destinations can be converted to and from.

Reply:
Yes. I also pointed to our architecture docs for more info - https://www.cloudquery.io/docs/developers/architecture